File: 2016/spring/cs152-04/homework/assignment02/object-oriented/binarysearch.java
import java.util.Random;

class BinaryLauncher {
    public static void main(String[] args) {

        Random random = new Random();
        int[] randInts = new int[10];

        System.out.println("Initilizing integer Array.");
        for (int i = 0; i < randInts.length; ++i) {
            randInts[i] = random.nextInt(101);
        }

        MergeSort mergeSort = new MergeSort(randInts);
        mergeSort.printArray();
        mergeSort.printSortedArray();
        
        BinarySearch binarySearch = new BinarySearch(mergeSort.getSortedArray());
        binarySearch.findValue(random.nextInt(101));
    }
}

class BinarySearch {

    BinarySearch(int[] sortedArray) {
        this.sortedArray = sortedArray;     
    }

    public void findValue(int value) {
        System.out.println("Querying Value: " + value);
        int index = findValue(this.sortedArray, 0, sortedArray.length-1, value);
        if (index >= 0) {
            System.out.println("Target value located at index " + index);
        }
        else {
            System.out.println("Target value does not exist in this array.");
        }
    }

    private int findValue(int[] sortedArray, int low, int high, int target) {
        while (low <= high) {
            int mid = (low + high) / 2;
            if (sortedArray[mid] > target) {
                high = mid-1; 
            }
            else if (sortedArray[mid] < target) {
                low = mid+1; 
            }
            else {
                return mid;
            }
        }
        return -1;
    }

    private int[] sortedArray;
}

class MergeSort {

    MergeSort(int[] intArr) {
        this.intArr = intArr;
        this.sortedArray = this.sort(intArr);
    }

    public void printArray() {
        System.out.println("Elements of the Array:");
        for (int i = 0; i < intArr.length-1; i++) {
            System.out.print(intArr[i] + ", ");        
        }
        System.out.println(intArr[intArr.length-1]);
    }

    public void printSortedArray() {
        System.out.println("Elements of the Sorted Array:");
        for (int i = 0; i < sortedArray.length-1; i++) {
            System.out.print(sortedArray[i] + ", ");
        }
        System.out.println(sortedArray[sortedArray.length-1]);
    }

    public int[] getSortedArray() {
        return sortedArray;
    }

    private int[] sort(int[] intArr) {
        int temp[] = new int[intArr.length];
        for (int i = 0; i < intArr.length; i++) {
            temp[i] = intArr[i];
        }

        sort(temp, 0, temp.length-1);
        return temp;
    }

    private void sort(int[] temp, int low, int high) {
        if (low < high) {
            int mid = (low + high) / 2;
            sort(temp, low, mid);
            sort(temp, mid+1, high);
            merge(temp, low, mid, high);
        }
    }

    private void merge(int[] temp, int low, int mid, int high) {
        int[] t = new int[temp.length];
        for (int i = low; i <= high; i++) {
           t[i] = temp[i]; 
        }

        int i = low;
        int j = mid + 1;
        int k = low;
        while (i <= mid && j <= high) {

            if (t[i] <= t[j]) {
                temp[k] = t[i];
                i++;
            }
            else {
                temp[k] = t[j];
                j++;
            }
            k++;
        }

        while (i <= mid) {
            temp[k] = t[i];
            k++;
            i++;
        }
    }

    private int[] intArr;
    private int[] sortedArray;
}

File: 2016/spring/cs152-04/homework/assignment02/procedural/binarysearch.c
#include <time.h>
#include <stdlib.h>
#include <stdio.h>
#define SWAP(a,b) { int t; t=a, a=b, b=t; }

int binarySearch(int*, int, int, int);
void mergeSort(int*, int, int);
void merge(int*, int, int, int);
void printIntArray(int*, int);

int main(void) {
    srand((unsigned) time(NULL)); 
    const int ARR_LEN = 10;
    int randInts[ARR_LEN];
    int targetElement, elementPosition;

    printf("Initializing array values.\n");
    int i;
    for (i = 0; i < ARR_LEN; ++i) {
        randInts[i] = rand() % 101;
    }
    mergeSort(randInts, 0, ARR_LEN-1);
    printIntArray(randInts, ARR_LEN);

    targetElement = rand() % 101;
    printf("\nInitiating binary search.\n");
    printf("Target element: %d\n", targetElement);

    elementPosition = binarySearch(randInts, 0, ARR_LEN-1, targetElement); 
    if (elementPosition >= 0) {
        printf("\nTarget element located at index %d\n", elementPosition);
    }
    else {
        printf("\nTarget element does not exist.\n");
    }
    return 0;
}

int binarySearch(int *intArr, int low, int high, int targetElement) {
    int mid;
    while (low <= high) {
       mid = (low + high) / 2; 
       if (intArr[mid] > targetElement) {
           high = mid - 1; 
       }
       else if (intArr[mid] < targetElement) {
           low = mid + 1;
       }
       else {
           return mid;
       }
    }
    return -1;
}

void printIntArray(int *intArr, int ARR_LEN) {
    int i;
    printf("%d", intArr[0]);
    for (i = 1; i < ARR_LEN; ++i) {
        printf(" %d", intArr[i]);
    }
}

void mergeSort(int *intArr, int low, int high) {
    int mid;
    if (low < high) {
        mid = (low + high) / 2;
        mergeSort(intArr, low, mid);
        mergeSort(intArr, mid+1, high);
        merge(intArr, low, mid, high);
    }
}

void merge(int *intArr, int low, int mid, int high) {
    int n1 = (mid - low) + 1;
    int n2 = (high - mid);
    int lTemp[n2], rTemp[n1];

    int i, j, k;
    for (i = 0; i < n1; i++) {
      lTemp[i] = intArr[low + i];
    }
    for (j = 0; j < n2; j++) {
      rTemp[j] = intArr[mid + j + 1];
    }

    i = 0;
    j = 0;
    k = low;
    while (i < n1 && j < n2) {
        if (lTemp[i] <= rTemp[j]) {
            intArr[k] = lTemp[i];
            i++;
        }
        else {
            intArr[k] = rTemp[j];
            j++;
        }
        k++;
    }

    while (i < n1) {
        intArr[k] = lTemp[i];
        i++;
        k++;
    }
    while (j < n2) {
        intArr[k] = rTemp[j];
        j++;
        k++;
    }
}
